Add support to a different unit of measurement(mmol/L) on Nightscout

marciogranzotto opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Bulding upon https://github.com/home-assistant/core/pull/54898 this PR uses the updated `py-nightscout` lib that has the new unit values conversion done, providing values like `sgv_mmol`  and `delta_mmol` . If the user selects the `mmol/L` unit, we just change which property of the response we use on the entity

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #45002
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. https://github.com/marciogranzotto/py-nightscout/releases
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mateusrevoredo committed:
Add support to a different unit of measurement on Nightscout

marciogranzotto committed:
Use new py-nightscout with mmol/L value

marciogranzotto committed:
Update nightscout delta attribute to use mmol if that's configured

frenck commented:
I don't think this needs to be an option flow. As a matter of fact, an option flow will cause problems as it can change the UoM for an entity, which isn't compatible with the long-term statistics.

marciogranzotto commented:
@frenck what you propose for people that need this value in `mmol/L` instead of `mg/dL`? I'm happy to give it a shot with some other kind of implementation

frenck commented:
Provide it as a secondary sensor (but disabled by default?)

marciogranzotto commented:
btw, the dexcom integration does this in the same way, on the config flow:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/dexcom/config_flow.py

frenck commented:
> btw, the dexcom integration does this in the same way, on the config flow:

I see, that will break stuff if the user changes it. So for that matter won't recommend on copying that pattern (nor will I approve such pattern).

marciogranzotto commented:
Yeah, I guess that makes sense with the long-term statistics.
I just don't really like the idea of having a disabled secondary sensor for that...
Maybe if we mode the unit or measurement to the actual setup and don't allow it to be changed afterwards?

frenck commented:
> Maybe if we move the unit or measurement to the actual setup and don't allow it to be changed afterwards?

I don't think that makes sense or is needed. There is nothing wrong with a secondary sensor.

marciogranzotto commented:
@frenck I think that the use case here is to always just use `mmol/L` or `mg/dL`.
Just to clarify for anyone that is now aware of how blood glucose works:
Different countries use different measurement units, much like the imperial/metric system. 
The international standard is `mmol/L`, but there's a lot of countries that use `mg/dL` such as the USA, Germany, and here in Brazil.
`mg/dL` is also the only unit that Nightscout exposes on their API. I think that's because it's an integer number, instead of a float on `mmol/L`.

Either way, my point is that I think the user should just choose which unit it's used where they live, on the setup flow, and never change it again since there's no need to.

frenck commented:
It could be part of the config flow (not option flow), that way it isn't changeable. But honestly, expose both sensors will satisfy any user. It isn't expensive.

markvader committed:
advise on use of an access token rather than API_SECRET

Proposed change to advise on use of an access token rather than API_SECRET, Provide link to nightscout guide.

markvader committed:
advise on use of an access token rather than API_SECRET

Proposed change to advise on use of an access token rather than API_SECRET, Provide link to nightscout guide.

markvader committed:
Merge pull request #1 from markvader/nightscout-strings-patch-1

advise on use of an access token rather than API_SECRET

markvader commented:
> `mg/dL` is also the only unit that Nightscout exposes on their API. I think that's because it's an integer number, instead of a float on `mmol/L`.
> 
> Either way, my point is that I think the user should just choose which unit it's used where they live, on the setup flow, and never change it again since there's no need to.

We've already made our selection into the unit of measurement to be used when we set up our personal nightscout sites and I agree this is unlikely to ever be changed.

@marciogranzotto The unit setting is exposed by the nightscout API (abet on the server_status call rather than individual entries/sgv calls).  
Can we use this call server_status.settings["units"] to set the particular setting that is enabled on sensor setup?

As a user i'd be happy to enable/disable the appropriate sensor if it was not set automatically

marciogranzotto commented:
@markvader thanks, I'll take a look on it!

bbredewold commented:
> @markvader thanks, I'll take a look on it!

Hi @marciogranzotto, did you have any time to look into this? I don't know if I could help, since I've never contributed to HA before... I'm very exciting use the mmol option soon ;)

marciogranzotto committed:
Merge pull request #2 from markvader/nightscout-strings-patch

Integration setup - advice on use of an access token rather than API_SECRET
